Geometry_Engine: Split bugs fixed and OutlinesFromLines method added

pawelbaran opened the pull request:
### NOTE: Depends on 
This PR will only work if merged into https://github.com/BHoM/BHoM_Engine/pull/3280 - let the latter get merged first to then get this one rebased.
   
### Issues addressed by this PR
<!-- Add reference(s) to issue(s) solved by this PR. Please use keyword Fixes/Closes as per https://help.github.com/articles/closing-issues-using-keywords/ -->

Closes #3260
Closes #3282
Closes #3284
Closes #3285
Closes #3289
Closes #3270 (@vietle-bh **please provide test file**)

<!-- Add short description of what has been fixed -->


### Test files
<!-- Link to test files to validate the proposed changes -->
On [SharePoint](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?RootFolder=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F02%5FPull%20Request%2FBHoM%2FBHoM%5FEngine%2FGeometry%5FEngine%2F%233283%2DOutlinesFromLines&FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4)


### Changelog
<!-- Text to go into changelog if applicable -->
<!-- Please see https://github.com/BHoM/documentation/wiki/changelog for guidelines -->


### Additional comments
<!-- As required -->
@FraserGreenroyd could you please have a look at `Split` as the original overviewer? It should be more robust now, edge cases ironed out
@michal-pekacki @adam-sobieski please review `OutlinesFromLines` - this is for your use mainly
@vietle-bh please test this PR against the issue you are having in #3270

I would like to skip UTs in this PR and do them all in one batch under https://github.com/BHoM/BHoM_Engine/issues/3277 once @peterjamesnugent is back, how does that sound @FraserGreenroyd?

FraserGreenroyd reviewed (changes requested):
Ok so I've reviewed the code and run a test and here be some feedback.

Comments on the code have been left at their relevant locations, I don't have any overarching comments regarding the algorithms of themselves at this time so no additional comments there.

On the test however, unfortunately, what's in this PR doesn't work for the original implementation of `Split` 😄 

![image](https://github.com/BHoM/BHoM_Engine/assets/18049174/59ab9352-f055-44c4-afe3-a9b452101ba6)

These are 2 models which split happily under the old code.

![image](https://github.com/BHoM/BHoM_Engine/assets/18049174/69235e43-e94b-41bd-a8ca-43055ace5271)

However on this branch I get this _helpful_ error message 😉 (see code comment for my comments on the UX of this error message 😋 ).

So on the basis of the test alone I'm afraid I need to request changes cause it no longer works but did previously 😄 I will upload my model to the linked folder as well for others.

pawelbaran commented on Geometry_Engine/Compute/OutlinesFromLines.cs:
```diff
@@ -0,0 +1,427 @@
+        {
+            Plane fitPlane = lines.SelectMany(x => x.ControlPoints()).ToList().FitPlane();
+            if (fitPlane == null)
+                return null;
```
The error is not raised on purpose - it is being thrown on the higher level, no need for stacking similar massages on top of each other in my opinion.

pawelbaran commented on Geometry_Engine/Compute/OutlinesFromLines.cs:
```diff
@@ -0,0 +1,427 @@
+            lines = lines?.CleanUpAndSplitWithEachOther(distanceTolerance);
+            if (lines == null)
+            {
+                BH.Engine.Base.Compute.RecordError("Outlines could not be created because the input set is not valid.");
```
The error is fully gone now after a slight change in code mechanics.

pawelbaran commented on Geometry_Engine/Compute/OutlinesFromLines.cs:
```diff
@@ -0,0 +1,427 @@
+        {
+            int initialCount = lines.Count;
+
+            Dictionary<Point, int> nodesByValence = new Dictionary<Point, int>();
```
Comment added 👍

pawelbaran commented on Geometry_Engine/Compute/OutlinesFromLines.cs:
```diff
@@ -0,0 +1,427 @@
+
+            foreach (Point point in nodesByValence.Keys.ToList())
+            {
+                if (nodesByValence.ContainsKey(point))
```
After the brief chat we had I removed this check without more thinking just to learn that it had been added on purpose 😃 It is needed because `CullValenceOne` method is recursive and may remove some keys from the `nodesByValence` dictionary. So actually it is an extremely important line of code.

pawelbaran commented on Geometry_Engine/Compute/OutlinesFromLines.cs:
```diff
@@ -0,0 +1,427 @@
+                    leftmost = node;
+            }
+
+            // Find edge starting at start node with leftmost dir
```
There is no Y check in this for loop and there had never been 😉 We are looking for the leftmost only point, purely X coordinate based exercise.

FraserGreenroyd replied:
Line 223 says `currentDir.Y < dir.Y` - should that say `X` then? I know the above for loop on line 210 was handling `X` but this for loop looks like it's handling `Y`?

pawelbaran replied:
Hmm, you may be right that I used a mental shortcut here while typing the code comment. Let's make sure we both are on the same page with the process itself:
1. We find leftmost node, i.e. a point lying at the left edge of the bounds
2. From that point, we find the direction looking leftmost in relation to the global Y - since all edges are on the right from the global Y (because we start from the leftmost node), the one with highest Y component will meet the description

So maybe `Find edge starting at leftmost start node with dir closest to global Y` would be more precise?

FraserGreenroyd replied:
I think that would make slightly more sense based on the code around the comment to me 👍

pawelbaran commented on Geometry_Engine/Compute/OutlinesFromLines.cs:
```diff
@@ -0,0 +1,427 @@
+                Vector newDir = null;
+                Line newEdge = null;
+
+                // To avoid computing heavy signed angle, first check only dirs to the left from the current dir
```
We are looking for directions to the left from the current direction, i.e. for edges in the left part of the space cut with infinite line defined by current node and current dir.

On an example, let's assume the green rectangle is our space (effectively bounds of points we process), blue is current node and dir. When we position ourselves at the node and look in the direction of dir, we first check for edges in orange area that meet at the node. Only if there is none of them, we check the white area. The reason for it, as in the code comment, is to avoid heave signed angle computations.

![image](https://github.com/BHoM/BHoM_Engine/assets/26874773/76c47480-5804-4a4f-a4a9-639051299615)

pawelbaran commented on Geometry_Engine/Compute/OutlinesFromLines.cs:
```diff
@@ -0,0 +1,427 @@
+            Point currentNode = leftmost;
+
+            // Look for subsequent edges with leftmost dirs compared to the dir of current edge
+            while (true)
```
Good point, I rewrote it to leverage `do/while` 👍

pawelbaran commented on Geometry_Engine/Compute/OutlinesFromLines.cs:
```diff
@@ -0,0 +1,427 @@
+
+                    if (newEdge == null)
+                    {
+                        BH.Engine.Base.Compute.RecordError("Outlines could not be created due to an internal error.");
```
It is a fallback, I would say this error should never be hit, but knowing the reality I put a generic message with return `null` to avoid a more serious crash.

pawelbaran commented:
Thanks for the review @FraserGreenroyd, all comments responded 👍

pawelbaran commented:
@BHoMBot check required

pawelbaran commented:
@BHoMBot check compliance
@BHoMBot check core
@BHoMBot check null-handling

pawelbaran commented:
@BHoMBot check compliance
@BHoMBot check core
@BHoMBot check null-handling

FraserGreenroyd reviewed (dismissed):
My test script is now working, I've reviewed the code and my comments have been addressed satisfactorily, so I'm happy to approve from my side @pawelbaran - happy to take another look if @michal-pekacki finds anything and we need some tweaks but until that moment, here is my approval for the `Split` element for the ceiling tile workflow 😄

pawelbaran commented:
Thanks @vietle-bh for the review, well spotted! I have raised #3289 to capture the issue, then fixed it in this PR, also included that case in test files linked to this PR.

Finally, I made use of your polygon to stress test the code, with satisfying results 😃 
![image](https://github.com/BHoM/BHoM_Engine/assets/26874773/f44ca87c-6a65-4a86-aa04-01faff7ec3bb)

pawelbaran commented:
@BHoMBot check required

pawelbaran commented:
One more edge case fix pushed + test added to the test files, the code should exactly as it did just more robust, all tests passing 👍

vietle-bh commented:
> Thanks @vietle-bh for the review, well spotted! I have raised #3289 to capture the issue, then fixed it in this PR, also included that case in test files linked to this PR.
> 
> Finally, I made use of your polygon to stress test the code, with satisfying results 😃 

I use the existing Split method intensively in a recursive method for a custom tool. The new Split method in this PR hasn't failed for the last several days, and seems to have resolved https://github.com/BHoM/BHoM_Engine/issues/3270 👍 

![image](https://github.com/BHoM/BHoM_Engine/assets/102604891/c305b7c5-7f7b-456a-aef7-1410f036feef)

vietle-bh commented:
Another great thing I got from this PR is a way to break down self-intersecting polylines. Thanks @pawelbaran 😍 
Out of interest, what did everyone previously use for breaking down self-intersecting polylines?

![image](https://github.com/BHoM/BHoM_Engine/assets/102604891/f04dc0d4-1569-4e98-8942-69b28b728a5a)

vietle-bh commented:
> Another great thing I got from this PR is a way to break down self-intersecting polylines. Thanks @pawelbaran 😍 Out of interest, what did everyone previously use for breaking down self-intersecting polylines?
> 
Guess what I'm really asking is which way is most performant 😋

vietle-bh commented:
Great teamwork! By the way, the Split method returns that generic "internal error" message if I feed it the self-intersecting line. It would be great to have it check for self-intersection and give a more specific message.

![image](https://github.com/BHoM/BHoM_Engine/assets/102604891/938ee38f-00e4-4b88-8237-d90fea406fdc)

pawelbaran commented:
> Great teamwork! By the way, the Split method returns that generic "internal error" message if I feed it the self-intersecting line. It would be great to have it check for self-intersection and give a more specific message.
> 
> ![image](https://private-user-images.githubusercontent.com/102604891/307084207-938ee38f-00e4-4b88-8237-d90fea406fdc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg2MjI0NzEsIm5iZiI6MTcwODYyMjE3MSwicGF0aCI6Ii8xMDI2MDQ4OTEvMzA3MDg0MjA3LTkzOGVlMzhmLTAwZTQtNGI4OC04MjM3LWQ5MGZlYTQwNmZkYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyMlQxNzE2MTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04MDg1YzkyYjA0ZTA5NjFmZWVmYjAyMTg0MjYyZjYyY2FkOTRlMzJlZDFiYWQwN2RhNTE4NzExMmQyNmE1MGYxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.omPmNLuJF62J3BeW2dHC4VNmE5rhGu6x_ERjbBdCO6I)

What do you mean by self-intersecting line? The outline?

pawelbaran commented:
I added more preliminary checks to the `Split` method, also included these cases in the test file @vietle-bh - should be all good now.

@FraserGreenroyd FYI I added `IsNull` overloads in BHoM_Engine and cleaned up the code a bit over there, could be worth you having a quick look 👍

michal-pekacki reviewed (dismissed):
Tested `OutlinesFromLines` and it works as it should, happy to approve 👍

pawelbaran commented:
@BHoMBot check required

pawelbaran commented:
@BHoMBot check required

michal-pekacki reviewed (approved):
Happy to re-approve 👍

pawelbaran commented:
@BHoMBot check copyright-compliance
@BHoMBot check dataset-compliance
@BHoMBot check unit-tests

FraserGreenroyd commented:
@BHoMBot this is a DevOps instruction. I am requesting neutral checks on: unit-tests

FraserGreenroyd commented:
@BHoMBot check ready-to-merge
